From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Matti=20Lehtim=C3=A4ki?= <matti.lehtimaki@gmail.com>
Date: Thu, 9 Feb 2017 21:48:19 +0200
Subject: [PATCH] (hybris) Disable SELinux
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Change-Id: I0511b2a0de1b20996f4fb8e9f3569acb41e2cf0f

(hybris) more SELinux disablement

Change-Id: Ic1bc474e993f2b66ab9114e88d7ec4f718a99d5c
Signed-off-by: Simonas Leleiva <simonas.leleiva@meramo.co.uk>

(hybris) disable remnants of SELinux which currently break

Change-Id: I252a8399c9853d993fc17d769bfbb2e4c722cfe8

(hybris) disable some more selinux functionality.

Change-Id: Ie63eb13b73d626668b32041fec4400cf2edd2fe5

(hybris) Disable SELinux init.

Sailfish OS SELinux initialization is done by systemd.
Conflicting init has been resolved by disabling SELinux
initialization from droid hal init.

Change-Id: I8a2a729671d224c4cc266a8219bee357611ded51
Signed-off-by: Matti Kosola <matti.kosola@jolla.com>
Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

(hybris) Disable unused code that was disabled by previous patches.

Change-Id: I2a78d91094495fad7b3cf514dd437d72a0718706
Signed-off-by: Björn Bidar <theodorstormgrade@gmail.com>

(hybris) allow second stage init previously disabled by selinux work

Signed-off-by: Simonas Leleiva <simonas.leleiva@jolla.com>
Change-Id: I1089428e88e7d94457f76745b23f44728dba8f6e

(hybris) execve service

Signed-off-by: Simonas Leleiva <simonas.leleiva@jolla.com>
Change-Id: I2efa4048bc2e4da848a61e24d8ef8b1824a81f06
---
 init/first_stage_init.cpp |  4 ++++
 init/init.cpp             | 11 +++++++++++
 init/property_service.cpp |  5 +++++
 init/selinux.cpp          | 16 +++++++++++++++-
 init/service.cpp          | 11 +++++++++++
 init/service_utils.cpp    |  2 ++
 init/subcontext.cpp       |  2 ++
 init/util.cpp             |  8 ++++++++
 8 files changed, 58 insertions(+), 1 deletion(-)

diff --git a/init/first_stage_init.cpp b/init/first_stage_init.cpp
index be0f373832aa094903fb7d2a2bae8aaee6866cd3..4a22afd7601060489f49c032a8bf798e4152c4e1 100644
--- a/init/first_stage_init.cpp
+++ b/init/first_stage_init.cpp
@@ -61,6 +61,8 @@ using namespace std::literals;
 
 namespace fs = std::filesystem;
 
+#define SYSTEMD_SELINUX
+
 namespace android {
 namespace init {
 
@@ -366,7 +368,9 @@ int FirstStageMain(int argc, char** argv) {
     gid_t groups[] = {AID_READPROC};
     CHECKCALL(setgroups(arraysize(groups), groups));
     //CHECKCALL(mount("sysfs", "/sys", "sysfs", 0, NULL));
+#ifndef SYSTEMD_SELINUX
     CHECKCALL(mount("selinuxfs", "/sys/fs/selinux", "selinuxfs", 0, NULL));
+#endif
 
     CHECKCALL(mknod("/dev/kmsg", S_IFCHR | 0600, makedev(1, 11)));
 
diff --git a/init/init.cpp b/init/init.cpp
index 4caf4f242d6449909127d0f8f358cec34073c4d5..63515c872ffa78f43bc9c07b2d93317a8db95fa6 100644
--- a/init/init.cpp
+++ b/init/init.cpp
@@ -119,6 +119,8 @@ using android::base::WriteStringToFile;
 using android::fs_mgr::AvbHandle;
 using android::snapshot::SnapshotManager;
 
+#define SYSTEMD_SELINUX
+
 namespace android {
 namespace init {
 
@@ -657,6 +659,7 @@ static Result<void> wait_for_coldboot_done_action(const BuiltinArguments& args)
 }
 
 static Result<void> SetupCgroupsAction(const BuiltinArguments&) {
+#if DISABLED_FOR_HYBRIS_SUPPORT
     if (!CgroupsAvailable()) {
         LOG(INFO) << "Cgroups support in kernel is not enabled";
         return {};
@@ -664,6 +667,7 @@ static Result<void> SetupCgroupsAction(const BuiltinArguments&) {
     if (!CgroupSetup()) {
         return ErrnoError() << "Failed to setup cgroups";
     }
+#endif
 
     return {};
 }
@@ -921,9 +925,12 @@ static void RecordStageBoottimes(const boot_clock::time_point& second_stage_star
 
     SetProperty("ro.boottime.init.first_stage",
                 std::to_string(selinux_start_time_ns - first_stage_start_time_ns));
+    // Mer: this seems to cause problems
+#if 0
     SetProperty("ro.boottime.init.selinux",
                 std::to_string(second_stage_start_time.time_since_epoch().count() -
                                selinux_start_time_ns));
+#endif
     if (auto init_module_time_str = getenv(kEnvInitModuleDurationMs); init_module_time_str) {
         SetProperty("ro.boottime.init.modules", init_module_time_str);
         unsetenv(kEnvInitModuleDurationMs);
@@ -1080,7 +1087,9 @@ int SecondStageMain(int argc, char** argv) {
     InitKernelLogging(argv);
     LOG(INFO) << "init second stage started!";
 
+#ifndef SYSTEMD_SELINUX
     SelinuxSetupKernelLogging();
+#endif
 
     // Update $PATH in the case the second stage init is newer than first stage init, where it is
     // first set.
@@ -1136,9 +1145,11 @@ int SecondStageMain(int argc, char** argv) {
     // Mount extra filesystems required during second stage init
     MountExtraFilesystems();
 
+#ifndef SYSTEMD_SELINUX
     // Now set up SELinux for second stage.
     SelabelInitialize();
     SelinuxRestoreContext();
+#endif
 
     Epoll epoll;
     if (auto result = epoll.Open(); !result.ok()) {
diff --git a/init/property_service.cpp b/init/property_service.cpp
index deb4ec805622761c25bd8b955c6926e83c192458..2c95126c1b71f8d1c205995aa3e542b81ebe6594 100644
--- a/init/property_service.cpp
+++ b/init/property_service.cpp
@@ -162,6 +162,8 @@ bool CanReadProperty(const std::string& source_context, const std::string& name)
 
 static bool CheckMacPerms(const std::string& name, const char* target_context,
                           const char* source_context, const ucred& cr) {
+    // Disable permission check in Mer
+    return true;
     if (!target_context || !source_context) {
         return false;
     }
@@ -277,12 +279,15 @@ class SocketConnection {
     }
 
     bool GetSourceContext(std::string* source_context) const {
+#if 0
         char* c_source_context = nullptr;
         if (getpeercon(socket_.get(), &c_source_context) != 0) {
             return false;
         }
         *source_context = c_source_context;
         freecon(c_source_context);
+#endif
+        *source_context = "";
         return true;
     }
 
diff --git a/init/selinux.cpp b/init/selinux.cpp
index 4baf9a4fe6cf71aae844ebd65ab6a9ee4979e521..1d369d46097652724b1c398402fdad247ef77f9a 100644
--- a/init/selinux.cpp
+++ b/init/selinux.cpp
@@ -93,6 +93,8 @@ using android::base::unique_fd;
 using android::fs_mgr::AvbHandle;
 using android::snapshot::SnapshotManager;
 
+#define SYSTEMD_SELINUX
+
 namespace android {
 namespace init {
 
@@ -100,6 +102,8 @@ namespace {
 
 enum EnforcingStatus { SELINUX_PERMISSIVE, SELINUX_ENFORCING };
 
+#if DISABLED_FOR_HYBRIS_SUPPORT
+
 EnforcingStatus StatusFromProperty() {
     std::string value;
     if (android::fs_mgr::GetKernelCmdline("androidboot.selinux", &value) && value == "permissive") {
@@ -117,6 +121,7 @@ bool IsEnforcing() {
     }
     return true;
 }
+#endif
 
 bool ReadFirstLine(const char* file, std::string* line) {
     line->clear();
@@ -130,6 +135,7 @@ bool ReadFirstLine(const char* file, std::string* line) {
     return true;
 }
 
+#if DISABLED_FOR_HYBRIS_SUPPORT
 Result<std::string> FindPrecompiledSplitPolicy() {
     std::string precompiled_sepolicy;
     // If there is an odm partition, precompiled_sepolicy will be in
@@ -183,6 +189,7 @@ Result<std::string> FindPrecompiledSplitPolicy() {
 
     return precompiled_sepolicy;
 }
+#endif
 
 bool GetVendorMappingVersion(std::string* plat_vers) {
     if (!ReadFirstLine("/vendor/etc/selinux/plat_sepolicy_vers.txt", plat_vers)) {
@@ -202,6 +209,7 @@ bool IsSplitPolicyDevice() {
     return access(plat_policy_cil_file, R_OK) != -1;
 }
 
+#if DISABLED_FOR_HYBRIS_SUPPORT
 std::optional<const char*> GetUserdebugPlatformPolicyFile() {
     // See if we need to load userdebug_plat_sepolicy.cil instead of plat_sepolicy.cil.
     const char* force_debuggable_env = getenv("INIT_FORCE_DEBUGGABLE");
@@ -439,6 +447,7 @@ void SelinuxSetEnforcement() {
         }
     }
 }
+#endif
 
 constexpr size_t kKlogMessageSize = 1024;
 
@@ -572,6 +581,7 @@ int SelinuxGetVendorAndroidVersion() {
     return vendor_android_version;
 }
 
+#ifndef SYSTEMD_SELINUX
 // This is for R system.img/system_ext.img to work on old vendor.img as system_ext.img
 // is introduced in R. We mount system_ext in second stage init because the first-stage
 // init in boot.img won't be updated in the system-only OTA scenario.
@@ -782,7 +792,10 @@ bool EarlySetupOverlays() {
 void SetupOverlays() {}
 #endif
 
+#endif
+
 int SetupSelinux(char** argv) {
+#ifndef SYSTEMD_SELINUX
     SetStdioToDevNull(argv);
     InitKernelLogging(argv);
 
@@ -828,8 +841,9 @@ int SetupSelinux(char** argv) {
     // SetupOverlays does not return if overlays exist, instead it execs overlay_remounter
     // which then execs second stage init
     if (use_overlays) SetupOverlays();
+#endif
 
-    const char* path = "/system/bin/init";
+    const char* path = "/sbin/droid-hal-init";
     const char* args[] = {path, "second_stage", nullptr};
     execv(path, const_cast<char**>(args));
 
diff --git a/init/service.cpp b/init/service.cpp
index e761285644ece20d726d83439a8492ea9a905612..9e9913374d69ae037015ca9bf062ab2cb1cc88b5 100644
--- a/init/service.cpp
+++ b/init/service.cpp
@@ -79,6 +79,9 @@ using android::base::WriteStringToFile;
 namespace android {
 namespace init {
 
+// Disable SELinux
+/*
+
 static Result<std::string> ComputeContextFromExecutable(const std::string& service_path) {
     std::string computed_context;
 
@@ -119,6 +122,7 @@ static Result<std::string> ComputeContextFromExecutable(const std::string& servi
     }
     return computed_context;
 }
+*/
 
 static bool ExpandArgsAndExecv(const std::vector<std::string>& args, bool sigstop) {
     std::vector<std::string> expanded_args;
@@ -262,11 +266,15 @@ void Service::SetProcessAttributesAndCaps(InterprocessFifo setsid_finished) {
         LOG(FATAL) << "cannot set attribute for " << name_ << ": " << result.error();
     }
 
+// Disable SELinux
+/*
+
     if (!seclabel_.empty()) {
         if (setexeccon(seclabel_.c_str()) < 0) {
             PLOG(FATAL) << "cannot setexeccon('" << seclabel_ << "') for " << name_;
         }
     }
+*/
 
     if (capabilities_) {
         if (!SetCapsForExec(*capabilities_)) {
@@ -672,6 +680,8 @@ Result<void> Service::Start() {
     }
 
     std::string scon;
+// Disable SELinux
+/*
     if (!seclabel_.empty()) {
         scon = seclabel_;
     } else {
@@ -681,6 +691,7 @@ Result<void> Service::Start() {
         }
         scon = *result;
     }
+*/
 
     if (!mount_namespace_.has_value()) {
         // remember from which mount namespace the service should start
diff --git a/init/service_utils.cpp b/init/service_utils.cpp
index 9343962459ed217cd9cb2b10ef810103a7277043..5db4104e5855ee6dcc88536bf2dafca21bdc23ac 100644
--- a/init/service_utils.cpp
+++ b/init/service_utils.cpp
@@ -313,6 +313,7 @@ Result<void> SetProcessAttributes(const ProcessAttributes& attr, InterprocessFif
     return {};
 }
 
+#if DISABLED_FOR_HYBRIS_SUPPORT
 Result<void> WritePidToFiles(std::vector<std::string>* files) {
     if (files->empty()) {
         // No files to write pid to, exit early.
@@ -368,6 +369,7 @@ Result<void> WritePidToFiles(std::vector<std::string>* files) {
     }
     return {};
 }
+#endif
 
 }  // namespace init
 }  // namespace android
diff --git a/init/subcontext.cpp b/init/subcontext.cpp
index 37f9fadeaf681535ec3c59bb16b2580937dd5f5b..4b005728868a058db8efa3b4b22359994653fa78 100644
--- a/init/subcontext.cpp
+++ b/init/subcontext.cpp
@@ -216,9 +216,11 @@ void Subcontext::Fork() {
         // We don't switch contexts if we're running the unit tests.  We don't use std::optional,
         // since we still need a real context string to pass to the builtin functions.
         if (context_ != kTestContext) {
+#if 0
             if (setexeccon(context_.c_str()) < 0) {
                 PLOG(FATAL) << "Could not set execcon for '" << context_ << "'";
             }
+#endif
         }
 #if defined(__ANDROID__)
         // subcontext init runs in "default" mount namespace
diff --git a/init/util.cpp b/init/util.cpp
index 6241b6617cb56b03b6b772da33ab0d9ead4d1331..134d6496f3d98a0bff0c23a1b5d6ed1c0b1a31ae 100644
--- a/init/util.cpp
+++ b/init/util.cpp
@@ -96,18 +96,22 @@ Result<uid_t> DecodeUid(const std::string& name) {
  */
 Result<int> CreateSocket(const std::string& name, int type, bool passcred, bool should_listen,
                          mode_t perm, uid_t uid, gid_t gid, const std::string& socketcon) {
+#if DISABLED_FOR_HYBRIS_SUPPORT
     if (!socketcon.empty()) {
         if (setsockcreatecon(socketcon.c_str()) == -1) {
             return ErrnoError() << "setsockcreatecon(\"" << socketcon << "\") failed";
         }
     }
+#endif
 
     android::base::unique_fd fd(socket(PF_UNIX, type, 0));
     if (fd < 0) {
         return ErrnoError() << "Failed to open socket '" << name << "'";
     }
 
+#if DISABLED_FOR_HYBRIS_SUPPORT
     if (!socketcon.empty()) setsockcreatecon(nullptr);
+#endif
 
     struct sockaddr_un addr;
     memset(&addr, 0 , sizeof(addr));
@@ -118,10 +122,12 @@ Result<int> CreateSocket(const std::string& name, int type, bool passcred, bool
         return ErrnoError() << "Failed to unlink old socket '" << name << "'";
     }
 
+#if 0
     std::string secontext;
     if (SelabelLookupFileContext(addr.sun_path, S_IFSOCK, &secontext) && !secontext.empty()) {
         setfscreatecon(secontext.c_str());
     }
+#endif
 
     if (passcred) {
         int on = 1;
@@ -133,9 +139,11 @@ Result<int> CreateSocket(const std::string& name, int type, bool passcred, bool
     int ret = bind(fd.get(), (struct sockaddr*)&addr, sizeof(addr));
     int savederrno = errno;
 
+#if 0
     if (!secontext.empty()) {
         setfscreatecon(nullptr);
     }
+#endif
 
     auto guard = android::base::make_scope_guard([&addr] { unlink(addr.sun_path); });
 
